### PR TITLE
Static pipeline column indexers, binary/regression evaluators

### LINKED ISSUE #868: Statically Piped Evaluators for Binary Classification and Regression
Now that we have a few trainers/scorers as part of the static pipelines, the next step might be evaluators. This is somewhat different than other components, since evaluators are not pipeline components (excepting per-instance evaluators, perhaps), as we see here in the aspirational examples.

https://github.com/dotnet/machinelearning/blob/df499aa9ea1d671e0df1c432622745c295e3e675/test/Microsoft.ML.Tests/Scenarios/Api/AspirationalExamples.cs#L133

One interesting part of the work is the introduction of a method of identifying columns. We already have this somewhat in the `.Append` method, where we are given an instance of the shape type, and select columns from it to show what to append. Here the technique looks identical, but we're in this case selecting a single item.

What makes me a little nervous about this approach is that if the operations are extension methods on the column types, I see absolutely no way to prevent someone from calling one of those extension methods inside that "indexer" delegate, short of adding more protections in the analyzer. (And, maybe 
this is fine, since the analyzer is clear and really this is probably a mistake a sensible person might plausibly make once.) More generally I talked about this in my remarks on ["context" here](https://github.com/dotnet/machinelearning/issues/632#issuecomment-416391466).

Related to #632.

### PR DESCRIPTION
**Pull request by TomFinley**
* Infrastructure for column indexers
* Binary classification and regression evaluators

Fixes #868.

### CODE REVIEW

File: src/Microsoft.ML.Data/StaticPipe/StaticPipeUtils.cs [hunk 1]
```diff
@@ -366,5 +367,97 @@ public BidirectionalDictionary()
...
+            /// <summary>
+            /// Constructor for the index helper. Note that any public or component code will instead use
+            /// <see cref="GetIndexer{T}(SchemaBearing{T})"/>, to fetch the lazily constructed instance of this
+            /// object, since its construction is somewhat expensive.
+            /// </summary>
+            /// <param name="schematized">Constructs the helper for this object.</param>
+            internal IndexHelper(SchemaBearing<T> schematized)
+            {
+                Indices = StaticPipeInternalUtils.MakeAnalysisInstance<T>(out var rec);
+                // We don't really care about this, but we need to work over something.
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/StaticPipe/StaticPipeUtils.cs]: >We don't really care about this, but we need to work over something. [](start = 19, length = 68)

I'm sorry, I completely don't understand what is going on. The comment adds negative clarity.

Comment (Zruty0) [reply src/Microsoft.ML.Data/StaticPipe/StaticPipeUtils.cs]: I see, you're constructing a method just to manufacture a `ParameterInfo`. Okay. Then you're calling `GetNamesValues`, but you only ever need values.

At least add a comment to `GetNamesValues` please

---
In reply to: [216415488](https://github.com/dotnet/machinelearning/pull/869#discussion_r216415488) [](ancestors = 216415488)

File: src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs [hunk 2]
```diff
@@ -787,6 +781,246 @@ private void ComputePrCurves()
...
             }
         }
+
+        /// <summary>
+        /// Evaluation results for binary classifiers, excluding probabilistic metrics.
+        /// </summary>
+        public class Result
+        {
+            /// <summary>
+            /// Gets the area under the ROC curve.
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs]: >Gets [](start = 16, length = 4)

not necessary on this PR, but maybe we can provide links to the longer explanation of the metrics (wikipedia or the MSFT doc on RML, maybe?), since those comments will be the user-facing documentation about the metrics (I assume)

File: src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs [hunk 3]
```diff
@@ -787,6 +781,246 @@ private void ComputePrCurves()
...
+
+            /// <summary>
+            /// Gets the area under the precision/recall curve of the classifier.
+            /// </summary>
+            /// <remarks>
+            /// The area under the precision/recall curve is a single number summary of the information in the
+            /// precision/recall curve. It is increasingly used in the machine learning community, particularly
+            /// for imbalanced datasets where one class is observed more frequently than the other. On these
+            /// datasets, AUPRC can highlight performance differences that are lost with AUC.
+            /// </remarks>
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs]: BTW, really like the explanation of the metric.

File: src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs [hunk 4]
```diff
@@ -787,6 +781,246 @@ private void ComputePrCurves()
...
+
+            /// <summary>
+            /// Gets the log-loss reduction (also known as relative log-loss, or reduction in information gain - RIG)
+            /// of the classifier.
+            /// </summary>
+            /// <remarks>
+            /// The log-loss reduction is scaled relative to a classifier that predicts the prior for every example:
+            /// (LL(prior) - LL(classifier)) / LL(prior)
+            /// This metric can be interpreted as the advantage of the classifier over a random prediction.
+            /// E.g., if the RIG equals 20, it can be interpreted as "the probability of a correct prediction is
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs]: >" [](start = 69, length = 1)

I think you might need to XML escape -> &quot;

File: src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs [hunk 5]
```diff
@@ -787,6 +781,246 @@ private void ComputePrCurves()
...
+        }
+
+        /// <summary>
+        /// Evaluates scored binary classification data.
+        /// </summary>
+        /// <typeparam name="T">The shape type for the input data.</typeparam>
+        /// <param name="data">The data to evaluate.</param>
+        /// <param name="label">The index delegate for the label column.</param>
+        /// <param name="pred">The index delegate for columns from calibrated prediction of a binary classifier.
+        /// Under typical scenarios, this will just be the same tuple of results returned from the trainer.</param>
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs]: >Under typical scenarios, this will just be the same tuple of results returned from the trainer [](start = 12, length = 94)

When I read this, I can't help thinking: under what scenarios this won't be the case.

File: src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs [hunk 6]
```diff
@@ -156,6 +158,129 @@ public override void AddColumn(ArrayDataViewBuilder dvBldr, string metricName, p
...
         }
+
+        public sealed class Result
+        {
+            /// <summary>
+            /// Gets the absolute loss of the model.
+            /// </summary>
+            /// <remarks>
+            /// The absolute loss is defined as
+            /// L1 = (1/m) * sum( abs( yi - y'i))
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs]: >' [](start = 45, length = 1)

 XML escape: &apos;

File: src/Microsoft.ML.Data/StaticPipe/StaticPipeUtils.cs [hunk 7]
```diff
@@ -13,7 +14,9 @@ namespace Microsoft.ML.Data.StaticPipe.Runtime
 {
     /// <summary>
     /// Utility methods for components that want to expose themselves in the idioms of the statically-typed pipelines.
-    /// These utilities are meant to be called by and useful to component authors, not users of those components.
+    /// These utilities are meant to be called by and useful to component authors, not users of those components. The
+    /// purpose is not to keep them hidden per se, but rather out of the face of users that are just trying to use the
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/StaticPipe/StaticPipeUtils.cs]: >out of the face of users [](start = 62, length = 24)

out of the attention of users

Comment (TomFinley) [reply src/Microsoft.ML.Data/StaticPipe/StaticPipeUtils.cs]: Yes I did not describe this well -- I think the flaw in my writing goes a bit beyond simple word choice. Let me rephrase this sentence entirely.

---
In reply to: [216475276](https://github.com/dotnet/machinelearning/pull/869#discussion_r216475276) [](ancestors = 216475276)

File: src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs (hunk 5 above)
Comment (TomFinley) [reply src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs]: Well, you could imagine someone composing their own dataset, not training anything or doing anything but loading a file. But I didn't feel the need to call this out.

---
In reply to: [216473441](https://github.com/dotnet/machinelearning/pull/869#discussion_r216473441) [](ancestors = 216473441)

File: src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs (hunk 6 above)
Comment (TomFinley) [reply src/Microsoft.ML.Data/Evaluators/RegressionEvaluator.cs]: OK. Though the comment does appear to render correctly from [the place where I copied it](https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.models.regressionmetrics.l1?view=ml-dotnet#Microsoft_ML_Models_RegressionMetrics_L1).

---
In reply to: [216474189](https://github.com/dotnet/machinelearning/pull/869#discussion_r216474189) [](ancestors = 216474189)

File: src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs (hunk 2 above)
Comment (TomFinley) [reply src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs]: I agree this might be helpful. Maybe we can review later, frankly I don't quite know how to insert links in XML docs since AFAIK `<a` is not a supported tag.

---
In reply to: [216470620](https://github.com/dotnet/machinelearning/pull/869#discussion_r216470620) [](ancestors = 216470620)

File: src/Microsoft.ML.Data/StaticPipe/StaticPipeUtils.cs (hunk 1 above)
Comment (TomFinley) [reply src/Microsoft.ML.Data/StaticPipe/StaticPipeUtils.cs]: Very well.

---
In reply to: [216416576](https://github.com/dotnet/machinelearning/pull/869#discussion_r216416576) [](ancestors = 216416576,216415488)
